src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,754 @@
+/***********************************************************************
+ * Copyright (c) 2021 Jonas Nick                                       *
+ * Distributed under the MIT software license, see the accompanying    *
+ * file COPYING or https://www.opensource.org/licenses/mit-license.php.*
+ ***********************************************************************/
+
+#ifndef SECP256K1_MODULE_MUSIG_SESSION_IMPL_H
+#define SECP256K1_MODULE_MUSIG_SESSION_IMPL_H
+
+#include <string.h>
+
+#include "../../../include/secp256k1.h"
+#include "../../../include/secp256k1_extrakeys.h"
+#include "../../../include/secp256k1_musig.h"
+
+#include "keyagg.h"
+#include "session.h"
+#include "../../eckey.h"
+#include "../../hash.h"
+#include "../../scalar.h"
+#include "../../util.h"
+
+static const unsigned char secp256k1_musig_secnonce_magic[4] = { 0x22, 0x0e, 0xdc, 0xf1 };
+
+static void secp256k1_musig_secnonce_save(secp256k1_musig_secnonce *secnonce, secp256k1_scalar *k) {
+    memcpy(&secnonce->data[0], secp256k1_musig_secnonce_magic, 4);
+    secp256k1_scalar_get_b32(&secnonce->data[4], &k[0]);
+    secp256k1_scalar_get_b32(&secnonce->data[36], &k[1]);
+}
+
+static int secp256k1_musig_secnonce_load(const secp256k1_context* ctx, secp256k1_scalar *k, secp256k1_musig_secnonce *secnonce) {
+    ARG_CHECK(secp256k1_memcmp_var(&secnonce->data[0], secp256k1_musig_secnonce_magic, 4) == 0);
+    secp256k1_scalar_set_b32(&k[0], &secnonce->data[4], NULL);
+    secp256k1_scalar_set_b32(&k[1], &secnonce->data[36], NULL);
+    return 1;
+}
+
+/* If flag is true, invalidate the secnonce; otherwise leave it. Constant-time. */
+static void secp256k1_musig_secnonce_invalidate(const secp256k1_context* ctx, secp256k1_musig_secnonce *secnonce, int flag) {
+    size_t i;
+    int zero = 0;
+    for (i = 0; i < sizeof(secnonce->data); i++) {
+        int val = secnonce->data[i];
+        secp256k1_int_cmov(&val, &zero, flag);
+        secnonce->data[i] = (unsigned char)val;
+    }
+    /* The flag argument is usually classified. So, above code makes the magic
+     * classified. However, we need the magic to be declassified to be able to
+     * compare it during secnonce_load. */
+    secp256k1_declassify(ctx, secnonce->data, sizeof(secp256k1_musig_secnonce_magic));
```

Comment:
Good to overwrite the magic. Maybe also add a (defense-in-depth) check in the load function that will abort if one  of the k is zero.

Reply:
I added this but I can't think of a scenario where this is useful. Should I remove it again?

Reply:
Maybe if the user accidentally overwrites part of the struct. Not very realistic I admit. 

I had a different aspect in mind: so far all of our "magic" checks are just sanity-checks. You could eliminate them from the code and (assuming a proper caller) nothing would change, and maybe this is what future devs/maintainers would implicitly assume about those checks. But this here was different, it has a real purpose in the API and removing it would lead to a bug.

I would keep it now, `scalar_is_zero` is cheap enough not too care.

Reply:
ok